keanu-project/src/main/java/io/improbable/keanu/util/DescriptionCreator.java

```diff
@@ -0,0 +1,178 @@
+        if (vertex instanceof ConstantVertex) {
+            Tensor value = vertex.getValue();
+            String valueString = value.isScalar() ? value.scalar().toString() :
+                new StringBuilder("(Tensor with shape: ").append(Arrays.toString(value.getShape())).append(")").toString();
+
+            return new StringBuilder()
+                .append("Const(")
+                .append(valueString)
+                .append(")")
+                .toString();
+        } else {
+            return "<<Unlabelled leaf Vertex>>";
+        }
+    }
+
+    static <T extends Tensor> String createDescriptionAllowingLabels(Vertex<T> vertex) {
+        Collection<Vertex> parents = vertex.getParents();
+
+        if (vertex.getLabel() != null) {
+            return vertex.getLabel().toString();
+        }
+
+        if (parents.size() == 0) {
+            return getBaseDescription(vertex);
+        }
+
+        return recursiveDescriptionStep(vertex, true);
+    }
+
+    private static <T extends Tensor> String recursiveDescriptionStep(Vertex<T> vertex, boolean includeBrackets) {
+        String irregularDescription = checkForIrregularExpressions(vertex, includeBrackets);
+        if (irregularDescription != null) {
+            return irregularDescription;
+        }
+
+        Stream<String> parentStream = vertex
+            .getParents()
+            .stream()
+            .map(DescriptionCreator::createDescriptionAllowingLabels);
+
+        String[] parentStrings = parentStream.toArray(String[]::new);
+
+        CharSequence delimiter = getDescriptionDelimiter(vertex);
+
+        StringBuilder builder = new StringBuilder();
+
+        if (includeBrackets) {
+            builder.append("(");
+        }
+        builder.append(String.join(delimiter, parentStrings));
+        if (includeBrackets) {
+            builder.append(")");
+        }
+
+        return builder.toString();
+    }
+
+    /**
+     * his method constructs an equation to describe how a vertex is calculated.
+     * The description is generated by recursively stepping up through the BayesNet and generating descriptions.
+     * Descriptions of common vertices will use infix operators.
+     * Descriptions will not recurse any further than labelled vertices.
+     *
+     * It is suggested that to use this feature, you label as many vertices as possible to avoid complex descriptions.
+     * @param <T> the type of the vertex
+     * @param vertex The vertex to create the description of
+     * @return An String equation describing how this vertex is calculated.<br>
+     * E.g. "This Vertex = that + (three * Const(4))"
+     */
+    public static <T extends Tensor> String createDescription(Vertex<T> vertex) {
+        Collection<Vertex> parents = vertex.getParents();
+
+        if (parents.size() == 0) {
+            if (vertex.getLabel() == null) {
+                return "<<Unlabelled Leaf Vertex>>";
```

Comment:
Probably best to try and limit the logic so you don't have to check this in multiple places, but if we really need this we should pull out the string to keep it consistent.